## PyliPD

This notebook serves as a test for using and evaluating the agent on PyliPD files through either local files or urls.

In [18]:
from pylipd.lipd import LiPD

In [21]:
D_remote = LiPD()
D_remote.set_endpoint("https://linkedearth.graphdb.mint.isi.edu/repositories/LiPDVerse-dynamic")
D_remote.load_remote_datasets(["Ocn-RedSea.Felis.2000"])
print("Loaded:", D_remote.get_all_dataset_names())

type(D_remote)

Caching datasets from remote endpoint..
Making remote query to endpoint: https://linkedearth.graphdb.mint.isi.edu/repositories/LiPDVerse-dynamic
Done..
Loaded: ['Ocn-RedSea.Felis.2000']


pylipd.lipd.LiPD

In [ ]:
D_url = LiPD()
D_url.load(["https://lipdverse.org/data/LCf20b99dfe8d78840ca60dfb1f832b9ec/1_0_1/Nunalleq.Ledger.2018.lpd"])
print("Loaded:", D_url.get_all_dataset_names())
D_url.get_timeseries_essentials()[["dataSetName", "paleoData_variableName", "paleoData_units"]]

In [20]:
data_path = 'Ocn-Palmyra.Nurhati.2011.lpd'
D = LiPD()
D.load(data_path)
type(D)

Loading 1 LiPD files


100%|██████████| 1/1 [00:00<00:00, 27.19it/s]

Loaded..


pylipd.lipd.LiPD

In [7]:
ts_list = D.get_timeseries(D.get_all_dataset_names())

type(ts_list)

Extracting timeseries from dataset: Ocn-Palmyra.Nurhati.2011 ...


dict

In [9]:
ts_list, df = D.get_timeseries(D.get_all_dataset_names(), to_dataframe=True)

df

Extracting timeseries from dataset: Ocn-Palmyra.Nurhati.2011 ...


,mode,time_id,archiveType,hasUrl,dataContributor,originalDataURL,lipdVersion,dataSetName,studyName,pub1_author,...,paleoData_values,paleoData_inferredVariableType,paleoData_dataType,paleoData_description,paleoData_inCompilation,paleoData_ocean2kID,paleoData_qCCertification,paleoData_iso2kUI,paleoData_pages2kID,paleoData_interpretation
0,paleoData,age,Coral,https://data.mint.isi.edu/files/lipd/Ocn-Palmy...,{'name': 'HLF MNE'},http://hurricane.ncdc.noaa.gov/pls/paleox/f?p=...,1.3,Ocn-Palmyra.Nurhati.2011,Nurhati2011_Palmyra,I.S. Nurhati,...,"[0.39, 0.35, 0.35, 0.35, 0.36, 0.22, 0.33, 0.3...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,paleoData,age,Coral,https://data.mint.isi.edu/files/lipd/Ocn-Palmy...,{'name': 'HLF MNE'},http://hurricane.ncdc.noaa.gov/pls/paleox/f?p=...,1.3,Ocn-Palmyra.Nurhati.2011,Nurhati2011_Palmyra,I.S. Nurhati,...,"[1998.21, 1998.13, 1998.04, 1997.96, 1997.88, ...",Year,float,Year AD,NaN,NaN,NaN,NaN,NaN,NaN
2,paleoData,age,Coral,https://data.mint.isi.edu/files/lipd/Ocn-Palmy...,{'name': 'HLF MNE'},http://hurricane.ncdc.noaa.gov/pls/paleox/f?p=...,1.3,Ocn-Palmyra.Nurhati.2011,Nurhati2011_Palmyra,I.S. Nurhati,...,"[1998.29, 1998.21, 1998.13, 1998.04, 1997.96, ...",Year,float,Year AD,NaN,NaN,NaN,NaN,NaN,NaN
3,paleoData,age,Coral,https://data.mint.isi.edu/files/lipd/Ocn-Palmy...,{'name': 'HLF MNE'},http://hurricane.ncdc.noaa.gov/pls/paleox/f?p=...,1.3,Ocn-Palmyra.Nurhati.2011,Nurhati2011_Palmyra,I.S. Nurhati,...,"[8.96, 8.9, 8.91, 8.94, 8.92, 8.89, 8.87, 8.81...",NaN,NaN,NaN,Ocean2k_v1.0.0,PacificNurhati2011,"MNE, NJA",CO11NUPM01BT1,Ocn_129,"[{'variableDetail': 'sea@surface', 'scope': 'c..."
4,paleoData,age,Coral,https://data.mint.isi.edu/files/lipd/Ocn-Palmy...,{'name': 'HLF MNE'},http://hurricane.ncdc.noaa.gov/pls/paleox/f?p=...,1.3,Ocn-Palmyra.Nurhati.2011,Nurhati2011_Palmyra,I.S. Nurhati,...,"[-5.41, -5.47, -5.49, -5.43, -5.48, -5.53, -5....",NaN,NaN,NaN,NaN,NaN,"MNE, NJA",CO11NUPM01B,NaN,"[{'variableDetail': 'sea_surface', 'direction'..."


In [10]:
df.columns

Index(['mode', 'time_id', 'archiveType', 'hasUrl', 'dataContributor',
       'originalDataURL', 'lipdVersion', 'dataSetName', 'studyName',
       'pub1_author', 'pub1_citeKey', 'pub1_title', 'pub1_institution',
       'pub1_url', 'pub1_urldate', 'pub2_author', 'pub2_year', 'pub2_citeKey',
       'pub2_volume', 'pub2_DOI', 'pub2_title', 'pub2_pages', 'pub2_journal',
       'pub2_dataUrl', 'pub2_publisher', 'pub3_author', 'pub3_issue',
       'pub3_citeKey', 'pub3_DOI', 'pub3_pages', 'pub3_title', 'pub3_year',
       'pub3_publisher', 'pub3_dataUrl', 'pub3_volume', 'pub3_journal',
       'googleDataURL', 'googleMetadataWorksheet', 'geo_meanLon',
       'geo_meanLat', 'geo_meanElev', 'geo_type', 'geo_ocean', 'geo_siteName',
       'geo_pages2kRegion', 'googleSpreadSheetKey', 'createdBy', 'tableType',
       'paleoData_measurementTableMD5', 'paleoData_measurementTableName',
       'paleoData_paleoDataTableName', 'paleoData_googleWorkSheetKey',
       'paleoData_filename', 'year', 'yearUnit

In [11]:
df['paleoData_variableName']

0     d18O
1     year
2     year
3    Sr/Ca
4     d18O
Name: paleoData_variableName, dtype: str

In [12]:
df[['paleoData_variableName','paleoData_notes']]

,paleoData_variableName,paleoData_notes
0,d18O,d18Osw (residuals calculated from coupled SrCa...
1,year,NaN
2,year,NaN
3,Sr/Ca,; paleoData_variableName changed - was origina...
4,d18O,Duplicate of modern d18O record presented in C...


In [13]:
df_cut = df.iloc[4,:]

df_cut

mode                                                                 paleoData
time_id                                                                    age
archiveType                                                              Coral
hasUrl                       https://data.mint.isi.edu/files/lipd/Ocn-Palmy...
dataContributor                                            {'name': 'HLF MNE'}
                                                   ...                        
paleoData_ocean2kID                                                        NaN
paleoData_qCCertification                                             MNE, NJA
paleoData_iso2kUI                                                  CO11NUPM01B
paleoData_pages2kID                                                        NaN
paleoData_interpretation     [{'variableDetail': 'sea_surface', 'direction'...
Name: 4, Length: 88, dtype: object

In [14]:
df_cut['paleoData_variableName']

'd18O'

In [15]:
df_essential = D.get_timeseries_essentials()

df_essential

,dataSetName,archiveType,geo_meanLat,geo_meanLon,geo_meanElev,paleoData_variableName,paleoData_values,paleoData_units,paleoData_proxy,paleoData_proxyGeneral,time_variableName,time_values,time_units,depth_variableName,depth_values,depth_units
0,Ocn-Palmyra.Nurhati.2011,Coral,5.87,-162.13,-10.0,d18O,"[-5.41, -5.47, -5.49, -5.43, -5.48, -5.53, -5....",permil,d18O,None,year,"[1998.29, 1998.21, 1998.13, 1998.04, 1997.96, ...",yr AD,None,None,None
1,Ocn-Palmyra.Nurhati.2011,Coral,5.87,-162.13,-10.0,Sr_Ca,"[8.96, 8.9, 8.91, 8.94, 8.92, 8.89, 8.87, 8.81...",mmol/mol,Sr/Ca,None,year,"[1998.29, 1998.21, 1998.13, 1998.04, 1997.96, ...",yr AD,None,None,None
2,Ocn-Palmyra.Nurhati.2011,Coral,5.87,-162.13,-10.0,d18O,"[0.39, 0.35, 0.35, 0.35, 0.36, 0.22, 0.33, 0.3...",permil,d18O,None,year,"[1998.21, 1998.13, 1998.04, 1997.96, 1997.88, ...",yr AD,None,None,None


In [16]:
df_essential.columns

Index(['dataSetName', 'archiveType', 'geo_meanLat', 'geo_meanLon',
       'geo_meanElev', 'paleoData_variableName', 'paleoData_values',
       'paleoData_units', 'paleoData_proxy', 'paleoData_proxyGeneral',
       'time_variableName', 'time_values', 'time_units', 'depth_variableName',
       'depth_values', 'depth_units'],
      dtype='str')

## Second Dataset Import
Can load directories besides invididual files

In [5]:
path = 'Pages2k/'

D_dir = LiPD()
D_dir.load_from_dir(path)

Loading 17 LiPD files


100%|██████████| 17/17 [00:00<00:00, 75.50it/s]

Loaded..


In [6]:
df_dir = D_dir.get_timeseries_essentials()

In [7]:
df_dir.head()

,dataSetName,archiveType,geo_meanLat,geo_meanLon,geo_meanElev,paleoData_variableName,paleoData_values,paleoData_units,paleoData_proxy,paleoData_proxyGeneral,time_variableName,time_values,time_units,depth_variableName,depth_values,depth_units
0,Ocn-RedSea.Felis.2000,Coral,27.8500,34.3200,-6.0,d18O,"[-4.12, -3.82, -3.05, -3.02, -3.62, -3.96, -3....",permil,d18O,None,year,"[1995.583, 1995.417, 1995.25, 1995.083, 1994.9...",yr AD,NaN,None,NaN
1,Ant-WAIS-Divide.Severinghaus.2012,Borehole,-79.4630,-112.1250,1766.0,temperature,"[-29.607, -29.607, -29.606, -29.606, -29.605, ...",degC,borehole,None,year,"[8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,...",yr AD,NaN,None,NaN
2,Ant-WAIS-Divide.Severinghaus.2012,Borehole,-79.4630,-112.1250,1766.0,uncertainty_temperature,"[1.327, 1.328, 1.328, 1.329, 1.33, 1.33, 1.331...",degC,NaN,None,year,"[8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,...",yr AD,NaN,None,NaN
3,Asi-SourthAndMiddleUrals.Demezhko.2007,Borehole,55.0000,59.5000,1900.0,temperature,"[0.166, 0.264, 0.354, 0.447, 0.538, 0.62, 0.68...",degC,borehole,None,year,"[800, 850, 900, 950, 1000, 1050, 1100, 1150, 1...",yr AD,NaN,None,NaN
4,Ocn-AlboranSea436B.Nieto-Moreno.2013,Marine sediment,36.2053,-4.3133,-1108.0,temperature,"[18.79, 19.38, 19.61, 18.88, 18.74, 19.25, 18....",degC,alkenone,None,year,"[1999.07, 1993.12, 1987.17, 1975.26, 1963.36, ...",yr AD,NaN,None,NaN


In [8]:
df_dir.shape

(28, 16)

In [9]:
df_dir['paleoData_variableName'].unique()

<StringArray>
[                   'd18O',             'temperature',
 'uncertainty_temperature',                   'notes',
                   'Mg_Ca',                   'Sr_Ca',
                    'Uk37',                   'trsgi',
                     'MXD']
Length: 9, dtype: str

In [10]:
df_temp = df_dir[df_dir['paleoData_variableName']=='temperature']
df_temp.head()

,dataSetName,archiveType,geo_meanLat,geo_meanLon,geo_meanElev,paleoData_variableName,paleoData_values,paleoData_units,paleoData_proxy,paleoData_proxyGeneral,time_variableName,time_values,time_units,depth_variableName,depth_values,depth_units
1,Ant-WAIS-Divide.Severinghaus.2012,Borehole,-79.4630,-112.1250,1766.0,temperature,"[-29.607, -29.607, -29.606, -29.606, -29.605, ...",degC,borehole,None,year,"[8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,...",yr AD,NaN,None,NaN
3,Asi-SourthAndMiddleUrals.Demezhko.2007,Borehole,55.0000,59.5000,1900.0,temperature,"[0.166, 0.264, 0.354, 0.447, 0.538, 0.62, 0.68...",degC,borehole,None,year,"[800, 850, 900, 950, 1000, 1050, 1100, 1150, 1...",yr AD,NaN,None,NaN
4,Ocn-AlboranSea436B.Nieto-Moreno.2013,Marine sediment,36.2053,-4.3133,-1108.0,temperature,"[18.79, 19.38, 19.61, 18.88, 18.74, 19.25, 18....",degC,alkenone,None,year,"[1999.07, 1993.12, 1987.17, 1975.26, 1963.36, ...",yr AD,NaN,None,NaN
8,Ocn-FeniDrift.Richter.2009,Marine sediment,55.5000,-13.9000,-2543.0,temperature,"[12.94, 10.99, 10.53, 10.44, 11.39, 13.38, 10....",degC,Mg/Ca,None,year,"[1998, 1987, 1975, 1962, 1949, 1936, 1924, 191...",yr AD,NaN,None,NaN
12,Ocn-FeniDrift.Richter.2009,Marine sediment,55.5000,-13.9000,-2543.0,temperature,"[12.94, 10.99, 10.53, 10.44, 11.39, 13.38, 10....",degC,NaN,None,year,"[1998, 1987, 1975, 1962, 1949, 1936, 1924, 191...",yr AD,depth_bottom,"[0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5, 7.5, 8.5, ...",cm


In [11]:
df_filt = df_dir.query("paleoData_variableName in ('temperature','MXD','Mg_Ca','d18O','trsgi', 'Uk37')")
df_filt.head()

,dataSetName,archiveType,geo_meanLat,geo_meanLon,geo_meanElev,paleoData_variableName,paleoData_values,paleoData_units,paleoData_proxy,paleoData_proxyGeneral,time_variableName,time_values,time_units,depth_variableName,depth_values,depth_units
0,Ocn-RedSea.Felis.2000,Coral,27.8500,34.3200,-6.0,d18O,"[-4.12, -3.82, -3.05, -3.02, -3.62, -3.96, -3....",permil,d18O,None,year,"[1995.583, 1995.417, 1995.25, 1995.083, 1994.9...",yr AD,NaN,None,NaN
1,Ant-WAIS-Divide.Severinghaus.2012,Borehole,-79.4630,-112.1250,1766.0,temperature,"[-29.607, -29.607, -29.606, -29.606, -29.605, ...",degC,borehole,None,year,"[8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,...",yr AD,NaN,None,NaN
3,Asi-SourthAndMiddleUrals.Demezhko.2007,Borehole,55.0000,59.5000,1900.0,temperature,"[0.166, 0.264, 0.354, 0.447, 0.538, 0.62, 0.68...",degC,borehole,None,year,"[800, 850, 900, 950, 1000, 1050, 1100, 1150, 1...",yr AD,NaN,None,NaN
4,Ocn-AlboranSea436B.Nieto-Moreno.2013,Marine sediment,36.2053,-4.3133,-1108.0,temperature,"[18.79, 19.38, 19.61, 18.88, 18.74, 19.25, 18....",degC,alkenone,None,year,"[1999.07, 1993.12, 1987.17, 1975.26, 1963.36, ...",yr AD,NaN,None,NaN
5,Eur-SpannagelCave.Mangini.2005,Speleothem,47.1000,11.6000,2347.0,d18O,"[-7.49, -7.41, -7.36, -7.15, -7.28, -6.99, -6....",permil,d18O,None,year,"[1935.0, 1932.0, 1930.0, 1929.0, 1929.0, 1928....",yr AD,NaN,None,NaN


In [13]:
D_dir.get_all_dataset_names()

['Ocn-RedSea.Felis.2000',
 'Ant-WAIS-Divide.Severinghaus.2012',
 'Asi-SourthAndMiddleUrals.Demezhko.2007',
 'Ocn-AlboranSea436B.Nieto-Moreno.2013',
 'Eur-SpannagelCave.Mangini.2005',
 'Ocn-FeniDrift.Richter.2009',
 'Eur-LakeSilvaplana.Trachsel.2010',
 'Ocn-Palmyra.Nurhati.2011',
 'Ocn-PedradeLume-CapeVerdeIslands.Moses.2006',
 'Ocn-SinaiPeninsula_RedSea.Moustafa.2000',
 'Eur-NorthernSpain.Martin-Chivelet.2011',
 'Arc-Kongressvatnet.D_Andrea.2012',
 'Eur-CoastofPortugal.Abrantes.2011',
 'Eur-SpanishPyrenees.Dorado-Linan.2012',
 'Eur-FinnishLakelands.Helama.2014',
 'Eur-NorthernScandinavia.Esper.2012',
 'Eur-Stockholm.Leijonhufvud.2009']

In [ ]:
D_src = LiPD()
D_src.set_endpoint("https://linkedearth.graphdb.mint.isi.edu/repositories/LiPDVerse-dynamic")
D_src.load_remote_datasets(["MD98_2181.Stott.2007", "Ocn-MadangLagoonPapuaNewGuinea.Kuhnert.2001"])

dses = D_src.get_datasets()
print(f"Got {len(dses)} Dataset objects, type: {type(dses[0])}")

D_reloaded = LiPD()
D_reloaded.load_datasets(dses)
print("Reloaded:", D_reloaded.get_all_dataset_names())

## Deduplication test
Merge the local Nurhati dataset (D) with the Pages2k directory (D_dir), which also contains Nurhati. The agent should only cite Ocn-Palmyra.Nurhati.2011 once.

In [ ]:
import pandas as pd

df_local = D.get_timeseries_essentials()
df_pages2k = D_dir.get_timeseries_essentials()

df_merged = pd.concat([df_local, df_pages2k], ignore_index=True)
print(f"Combined rows: {len(df_merged)}")
print(f"Unique datasets: {df_merged['dataSetName'].nunique()}")
print(df_merged['dataSetName'].unique())

In [14]:
bibs, df = D_dir.get_bibtex(remote = True, save = True, path = 'mybiblio.bib', verbose = True)

Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (nan), creating the entry manually
Cannot find a matching record for the provided DOI (na

In [15]:
bibs

['@misc{felis2000httpswwwncdcnoaagovpaleostudy1861DataCitation,\n    author = "T. Felis",\n    title = "World Data Center for Paleoclimatology",\n    institution = "World Data Center for Paleoclimatology",\n    url = "https://www.ncdc.noaa.gov/paleo/study/1861"\n}\n',
 ' @article{Felis_2000, title={A coral oxygen isotope record from the northern Red Sea documenting NAO, ENSO, and North Pacific teleconnections on Middle East climate variability since the year 1750}, volume={15}, ISSN={1944-9186}, url={http://dx.doi.org/10.1029/1999PA000477}, DOI={10.1029/1999pa000477}, number={6}, journal={Paleoceanography}, publisher={American Geophysical Union (AGU)}, author={Felis, Thomas and Pätzold, Jürgen and Loya, Yossi and Fine, Maoz and Nawar, Ahmed H. and Wefer, Gerold}, year={2000}, month=Dec, pages={679–694} }\n',
 ' @article{Tierney_2015, title={Tropical sea surface temperatures for the past four centuries reconstructed from coral archives}, volume={30}, ISSN={1944-9186}, url={http://dx.doi

In [17]:
df.head()

,dsname,title,authors,doi,pubyear,year,journal,volume,issue,pages,type,publisher,report,citeKey,edition,institution,url,url2
0,Ocn-RedSea.Felis.2000,World Data Center for Paleoclimatology,T. Felis,NaN,None,NaN,NaN,NaN,None,NaN,dataCitation,NaN,None,felis2000httpswwwncdcnoaagovpaleostudy1861Data...,None,World Data Center for Paleoclimatology,None,https://www.ncdc.noaa.gov/paleo/study/1861
1,Ocn-RedSea.Felis.2000,A coral oxygen isotope record from the norther...,Jürgen Pätzold and Yossi Loya and Thomas Felis...,10.1029/1999PA000477,None,2000.0,Paleoceanography,15,6.0,679-694,article,Wiley-Blackwell,None,felis2000acoraloxygenisotoperecord,None,NaN,None,NaN
2,Ocn-RedSea.Felis.2000,Tropical sea surface temperatures for the past...,Kevin J. Anchukaitis and Michael N. Evans and ...,10.1002/2014PA002717,None,2015.0,Paleoceanography,30,3.0,226-252,journal-article,Wiley-Blackwell,None,tierney2015tropicalseasurfacetempera,None,NaN,None,NaN
3,Ant-WAIS-Divide.Severinghaus.2012,This study,Severinghaus J.,NaN,None,NaN,NaN,NaN,None,NaN,dataCitation,NaN,None,severinghaus0thisstudyDataCitation,None,NaN,None,This study
4,Ant-WAIS-Divide.Severinghaus.2012,Little Ice Age cold interval in West Antarctic...,Bruce D. Cornuelle and Jeffrey P. Severinghaus...,10.1029/2012GL051260,None,2012.0,Geophysical Research Letters,39,9.0,L09710,article,Wiley-Blackwell,None,orsi2012littleiceagecoldintervali,None,NaN,None,NaN


## Datasets that should be cited

This notebook uses the following datasets, which should appear in a generated bibliography:

- **Ocn-RedSea.Felis.2000** — Felis, T. et al. (2000). Coral d18O from the northern Red Sea. *Paleoceanography*, 15(6), 679–694. doi:10.1029/1999PA000477
- **Nunalleq.Ledger.2018** — Ledger, P. M. et al. (2018). Climate reconstruction from archaeological site Nunalleq, Alaska.
- **Ocn-Palmyra.Nurhati.2011** — Nurhati, I. S., Cobb, K. M., & Di Lorenzo, E. (2011). Decadal-scale SST and salinity variations in the central tropical Pacific. *Journal of Climate*, 24(13), 3294–3308. doi:10.1175/2011JCLI3852.1
- **Pages2k directory** (17 datasets) — including Felis.2000, Severinghaus.2012, Demezhko.2007, Nieto-Moreno.2013, Mangini.2005, Richter.2009, Trachsel.2010, Nurhati.2011, Moses.2006, Moustafa.2000, Martin-Chivelet.2011, D_Andrea.2012, Abrantes.2011, Dorado-Linan.2012, Helama.2014, Esper.2012, Leijonhufvud.2009
- **MD98_2181.Stott.2007** — Stott, L. et al. (2007). Deep sea sediment core MD98-2181 from the western Pacific warm pool.
- **Ocn-MadangLagoonPapuaNewGuinea.Kuhnert.2001** — Kuhnert, H. et al. (2001). Coral record from Madang Lagoon, Papua New Guinea.